### PR TITLE
Color theme

### PR DESCRIPTION
**Pull request by ignaspa**
## Motivation and Context

In the pursuit of increased color accessibility and theme-ing, we are trying to introduce a darkmode.

## Your Changes

**Description**:
Creating variable based CSS coloring and styling in the SCSS files and adding settings for users to change the UI for their preference.

**Type of change** (select all that apply):

 

- [x] New feature (non-breaking change which adds functionality)

## Testing

Checking interface changes on firefox and chrome.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have updated the Changelog.md file.

### CODE REVIEW

File: app/assets/stylesheets/exam_templates.scss [hunk 1]
```diff
@@ -50,14 +50,14 @@
...
-        background-color: $lighter-blue;
+        background-color: $background-support;
       }
     }
   }
 
   .add_template {
     margin-top: 10px;
-    background-color: $light-blue !important;
+    background-color: $background-support !important;
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/exam_templates.scss:60]: You can delete this `!important`

File: app/assets/stylesheets/grader.scss [hunk 2]
```diff
@@ -51,32 +51,32 @@
...
+    border-bottom: 1px solid $background-main;
     padding: 5px;
     width: 100%;
     display: inline-block;
   }
 
   .unassigned {
-    background: $light-grey;
-    color: $grey;
+    background: $background-main;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/grader.scss:61]: This one is supposed to be faded to indicate that it's read-only.

File: app/assets/stylesheets/grader.scss [hunk 3]
```diff
@@ -51,32 +51,32 @@
   width: 100%;
 
   li {
-    border-bottom: 1px solid $light-grey;
+    border-bottom: 1px solid $background-main;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/grader.scss:54]: Won't this make the border blend in with the background?

File: app/assets/stylesheets/common/_constants.scss [hunk 4]
```diff
@@ -1,29 +1,63 @@
...
+}
+// Code viewer
+$comments: var(--comments_color);
+$key-words: var(--key_words_color);
+$strings: var(--strings_color);
+
+// General contexts
+$links: var(--primary_one);
+$pane-highlight: var(--primary_two);
+$confirmed: var(--light_success);
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_constants.scss:31]: This is only used in two places. For the selected rubric level, maybe let's use one of the `primary` variables? And for the other place (success flash), let's extend the pattern you have a few lines down and introduce `$success` and `$success-light`.

File: app/assets/stylesheets/groups.scss [hunk 5]
```diff
@@ -30,7 +30,7 @@
 }
 
 .display_pane {
-  border: 1px solid $light-grey;
+  border: 1px solid $background-main;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/groups.scss:33]: Similar to a previous comment, doesn't this cause the border to blend into the background? (I won't comment at other places I see this.)

File: app/javascript/packs/application.js [hunk 6]
```diff
@@ -53,13 +53,18 @@ import 'javascripts/layouts';
 import 'javascripts/menu';
 import 'javascripts/react_config';
 import 'javascripts/redirect';
+import 'javascripts/theme_colors';
```
Comment (david-yz-liu) [inline app/javascript/packs/application.js:56]: You can remove this

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -3,6 +3,8 @@
     <strong><%= Rails.configuration.course_name %></strong>
   </div>
   <div id='user-info'>
+    <input type="checkbox" id="vehicle3" name="vehicle3" value="Boat" onclick="switch_theme()">
```
Comment (david-yz-liu) [inline app/views/layouts/_header.html.erb:6]: You can remove these two lines (and remove the `switch_theme` function).

File: app/views/students/mailer_settings.html.erb [hunk 8]
```diff
@@ -8,6 +8,10 @@
     <%= f.check_box :receives_invite_emails %>
     <%= f.label :receives_invite_emails, t('students.toggle_invite_notifications') %>
   </p>
+  <p>
+    <%= f.check_box :theme, {}, 'dark', 'light' %>
```
Comment (david-yz-liu) [inline app/views/students/mailer_settings.html.erb:12]: Let's make this a `select`

File: config/locales/views/users/en.yml [hunk 9]
```diff
@@ -26,6 +26,7 @@ en:
     not_successfully_added_message_5: "Each line has the required fields specified near the upload form."
     toggle_invite_notifications: 'Receive email notifications when I am invited to a group.'
     toggle_results_notifications: 'Receive email notifications when results are released.'
+    toggle_theme: 'Enable Dark Theme.'
```
Comment (david-yz-liu) [inline config/locales/views/users/en.yml:29]: Let's call this `Color theme` (since it'll be paired with a `select` input)

File: db/migrate/20201029164928_add_theme_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddThemeToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :theme, :string, default: 'light', null: false
```
Comment (david-yz-liu) [inline db/migrate/20201029164928_add_theme_to_users.rb:3]: Let's make this new column an Enum: https://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html.

File: app/views/assignments/_boot.js.erb [hunk 11]
```diff
@@ -45,7 +45,7 @@
     $('.assignment_file').each(function() {
       if (!filename_re.exec(this.value)) {
         invalid = true;
-        // _constants.scss $red
+        // _constants.scss $severe-error
         $(this).css('background-color', '#fe2a2a');
```
Comment (david-yz-liu) [inline app/views/assignments/_boot.js.erb:49]: I think you can set this to a `--var(...)`? In which case you can also remove the comment.

File: app/assets/stylesheets/submissions.scss [hunk 12]
```diff
@@ -19,7 +19,7 @@
 }
 
 .submission_collected {
-  background-color: $light-green !important;
+  background-color: $confirmed !important;
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/submissions.scss:22]: You can remove this `!important`, and the one for `.submission_error` below.

File: app/assets/stylesheets/submissions.scss [hunk 13]
```diff
@@ -62,12 +62,12 @@
       visibility: hidden;
 
       li {
-        color: $dark-blue !important;
+        color: $primary-one !important;
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/submissions.scss:65]: This one and the ones down below can be ignored (I'm pretty sure we can remove this CSS altogether, I'll review more carefully later).

File: app/assets/stylesheets/common/_constants.scss [hunk 14]
```diff
@@ -1,29 +1,63 @@
...
+$annotation-holder: var(--annotation_holder);
+
+// Status
+$severe-error: var(--heavy_error);
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_constants.scss:35]: Dividing into sections is good; within each section, order variables alphabetically.

File: app/assets/stylesheets/common/_markus.scss [hunk 15]
```diff
@@ -79,18 +79,19 @@ p + p {
...
   &.inactive {
-    color: $black;
+    color: $sharp-line;
     pointer-events: none;
   }
 
   &:hover {
-    color: $blue;
+    color: $sharp-line;
+    font-weight: bold;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:94]: Remove this (we can discuss accessibility for links later)

File: app/assets/stylesheets/common/_markus.scss [hunk 16]
```diff
@@ -79,18 +79,19 @@ p + p {
...
 
   &.inactive {
-    color: $black;
+    color: $sharp-line;
     pointer-events: none;
   }
 
   &:hover {
-    color: $blue;
+    color: $sharp-line;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:93]: actually you can remove this too

File: app/assets/stylesheets/common/_markus.scss [hunk 17]
```diff
@@ -79,18 +79,19 @@ p + p {
 /** Links */
 
 a {
-  color: $dark-blue;
+  color: $primary-one;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:82]: I think this should be `$links`?

File: app/assets/stylesheets/common/_markus.scss [hunk 18]
```diff
@@ -179,15 +180,15 @@ textarea {
   }
 
   &[disabled] {
-    background-color: $disabled-grey;
-    border-color: $disabled-dark;
-    color: $disabled-dark;
+    background-color: $background-support;
+    border-color: $background-support;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:184]: I think this one does need a different colour, otherwise the border will disappear

File: app/assets/stylesheets/common/_markus.scss [hunk 19]
```diff
@@ -350,25 +352,25 @@ fieldset {
 }
 
 .disable {
-  color: $dark-grey;
+  color: $line;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:355]: I just thought about this here, but I think we should probably have a constant `$disabled` or something, so that it's not strictly tied to line colour. (Would be worth reviewing all the colours that represent "disabled" elements.)

File: app/assets/stylesheets/common/_markus.scss [hunk 20]
```diff
@@ -658,16 +660,16 @@ ul.tags {
...
         cursor: pointer;
         display: inline-flex;
         list-style: none;
         padding: 0.5em;
         width: 100%;
 
         &:hover {
-          background: $blue;
-          color: $white;
+          background: $background_support;
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/common/_markus.scss:671]: Looks like there's a typo here

File: app/assets/stylesheets/common/_markus.scss [hunk 21]
```diff
@@ -948,24 +950,25 @@ nav {
...
   position: absolute;
   z-index: 1;
 }
 
 .annotation-holder-active {
-  background: $orange-transparent;
-  border: 3px solid $red;
+  background: $annotation-holder;
+  border: 3px solid $severe-error;
+  opacity: 0.2;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:963]: It's okay to use the same colour as `.annotation_holder`, but let's increase the opacity a little to make this stand out.

**Review by david-yz-liu — COMMENTED** [review]
Nice work! Left some preliminary comments

File: app/assets/stylesheets/common/_constants.scss [hunk 22]
```diff
@@ -1,29 +1,63 @@
+:root {
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/assets/stylesheets/common/_constants.scss:1]: @david-yz-liu Is this something I can avoid?

File: app/assets/stylesheets/grader.scss (hunk 3 above)
Comment (ignaspa) [reply app/assets/stylesheets/grader.scss:54]: Yes, whoops!

File: app/assets/stylesheets/grader.scss (hunk 2 above)
Comment (ignaspa) [reply app/assets/stylesheets/grader.scss:61]: Is this:
image
more in line with what the disabled marks should look like?

File: app/assets/stylesheets/groups.scss (hunk 5 above)
Comment (ignaspa) [reply app/assets/stylesheets/groups.scss:33]: Definitely a lot of places with this, hopefully I'm catching them all!

File: app/assets/stylesheets/common/SyntaxHighlighter.scss [hunk 23]
```diff
@@ -130,15 +130,15 @@
   }
 
   .footer {
-    background-color: #eceadb;
-    border-top: 1px solid #fff;
-    color: #333;
+    background-color: $background-support;
+    border-top: 1px solid $background-main;
```
Comment (ignaspa) [inline app/assets/stylesheets/common/SyntaxHighlighter.scss:134]: I'm not too sure where this is being used. The dp-about class this is nested within is only only used in shCore.js.

File: app/assets/stylesheets/groups.scss (hunk 5 above)
Comment (ignaspa) [reply app/assets/stylesheets/groups.scss:33]: **Fixed**:
Line 126 of exam_templates.scss is caught and fixed.
Line 140 of exam_templates.scss is caught and fixed.

**Questions**:
Line 133 of SyntaxHighlighter.scss has a similar occurrence but I am confused about where that applies. I left a comment on that file.

Lines 33 and 46 in groups.scss are essentially the same as the ones in exam_templates.scss, yet I cannot find where they are used for graders in either assignment settings or the graders list in users. All of the lines were authored in the same commit by Han in revision 1736b25. I think I should just removed these classes in the groups.scss file, since they are defined in the exam_templates.scss anyway. Can I do so?

**No fix needed**:
Line 795 of _markus.scss has a similar occurrence which is intentional and correct.
Line 462 of _navigation.scss has a similar occurrence which is intentional and correct.

File: app/views/assignments/_boot.js.erb (hunk 11 above)
Comment (ignaspa) [reply app/views/assignments/_boot.js.erb:49]: Ok, I made the change but I'm not too sure what `check_required_fields()` does

File: app/assets/stylesheets/common/_constants.scss (hunk 22 above)
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/common/_constants.scss:1]: No you can resolve this one

File: app/assets/stylesheets/grader.scss (hunk 2 above)
Comment (david-yz-liu) [reply app/assets/stylesheets/grader.scss]: I like that more, although I would prefer that the "selected level" colour be different than the "disabled" colour

File: app/views/assignments/_boot.js.erb (hunk 11 above)
Comment (david-yz-liu) [reply app/views/assignments/_boot.js.erb]: It's doing some client-side validation for the assignments form (enabling/disabling the "Submit" button)

File: app/assets/stylesheets/common/SyntaxHighlighter.scss (hunk 23 above)
Comment (david-yz-liu) [reply app/assets/stylesheets/common/SyntaxHighlighter.scss:133]: You should change it. I'll investigate later, it might be that we can remove this whole `.dp-about` ruleset.

File: app/assets/javascripts/Components/Helpers/data_chart.jsx [hunk 24]
```diff
@@ -15,9 +15,12 @@ export class DataChart extends React.Component {
       responsive: false,
       legend: {
         display: this.props.legend
-      },
+        },
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Helpers/data_chart.jsx:18]: Revert this change

File: app/assets/javascripts/Components/Helpers/data_chart.jsx [hunk 25]
```diff
@@ -29,14 +32,16 @@ export class DataChart extends React.Component {
...
+            color: document.documentElement.style.getPropertyValue('--primary_three')
+          },
           scaleLabel: {
             display: true,
             labelString: this.props.xTitle
           }
         }]
       }
     };
-
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Helpers/data_chart.jsx:39]: Revert this change

File: app/assets/javascripts/chart_config.js [hunk 26]
```diff
@@ -1,17 +1,45 @@
...
-  barPercentage: 0.9,
-  categoryPercentage: 0.8
-}];
+  Chart.defaults.global.type = 'bar';
+
+  Chart.defaults.bar.scales.xAxes = [{
+    ticks: {
+      type: 'linear',
+      fontColor: ticksColor,
+
```
Comment (david-yz-liu) [inline app/assets/javascripts/chart_config.js:29]: Delete this line

File: app/assets/stylesheets/assignments.scss [hunk 27]
```diff
@@ -80,6 +80,24 @@ fieldset {
 }
 
 
+/* Grader view */
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/assignments.scss:83]: I deleted this code in https://github.com/MarkUsProject/Markus/pull/4967/files, you can just remove it.

File: app/assets/stylesheets/common/_core.scss [hunk 28]
```diff
@@ -237,18 +237,43 @@ label.required::after, th.required::after {
...
+  }
+
+  .rt-thead .rt-th, .rt-thead .rt-td {
+    color: $line;
+  }
+
   .rt-td:last-child {
     // Hack to ensure scrollbar doesn't cover cell contents.
     padding-right: 22px;
+    border-right: 1px solid $line;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_core.scss:275]: I'm not sure why these two rules are necessary, since this should just apply to the last cell.

File: config/locales/views/users/en.yml [hunk 29]
```diff
@@ -27,6 +27,7 @@ en:
     toggle_invite_notifications: 'Receive email notifications when I am invited to a group.'
     toggle_results_notifications: 'Receive email notifications when results are released.'
     display_name: 'Display name'
+    toggle_theme: 'Color Theme'
```
Comment (david-yz-liu) [inline config/locales/views/users/en.yml:30]: This shouldn't be here, it should be under `models/users` as an attribute name

File: app/assets/stylesheets/common/_react_tabs.scss [hunk 30]
```diff
@@ -10,6 +12,7 @@
 
 .react-tabs__tab-panel {
   flex-grow: 1;
+  scrollbar-color: $line $background-main;
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/assets/stylesheets/common/_react_tabs.scss:15]: @david-yz-liu this is an emergent property of css, currently only supported by firefox. Eventually, we can remove the webkit scrollbar styling and simply have this.

File: app/assets/stylesheets/common/_core.scss [hunk 31]
```diff
@@ -235,17 +236,85 @@ label.required::after, th.required::after {
...
+    border-bottom: 1px solid $line;
+    border-right: 1px solid $line;
+  }
+
+  .rt-tbody .rt-td {
+    border-bottom: 1px solid $line;
+    border-right: 1px solid $line;
+  }
+
+  .rt-thead .rt-th,
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/assets/stylesheets/common/_core.scss:311]: @david-yz-liu these selectors are copied from how react table applies them, does this hound comment mean I need to do the `& >` syntax like one of the following two equivalencies?

1)

image

2)
image

[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/assets/stylesheets/common/_core.scss:311]: You can ignore this warning.
